File: 2755-extra-characters-in-a-string/2755-extra-characters-in-a-string.js
/**
 * @param {string} s
 * @param {string[]} dictionary
 * @return {number}
 */
function dfs(s, dp, wordSet, idx) {
    if (idx >= s.length) return 0;
    if (dp[idx] !== -1) return dp[idx];

    let minlen = s.length;
    let currStr = "";
    for (let i = idx; i < s.length; i++) {
        currStr += s[i];
        let count = wordSet.has(currStr) ? 0 : currStr.length;
        let newCount = dfs(s, dp, wordSet, i + 1);
        minlen = Math.min(minlen, count + newCount);
    }

    dp[idx] = minlen;
    return dp[idx];
}

var minExtraChar = function(s, dictionary) {
    let dp = Array(s.length).fill(-1);
    let wordSet = new Set(dictionary);  
    return dfs(s, dp, wordSet, 0);
};

File: 2755-extra-characters-in-a-string/2755-extra-characters-in-a-string.py
class Solution(object):
    def dfs(self, s, dp, st, idx):
        if idx >= len(s):
            return 0
        if dp[idx] != -1:
            return dp[idx]
        
        minlen, _str = len(s), ""
        for i in range(idx, len(s)):
            _str += s[i]
            count = 0 if _str in st else len(_str)
            new_count = self.dfs(s, dp, st, i + 1)
            minlen = min(minlen, count + new_count)
        
        dp[idx] = minlen
        return dp[idx]

    def minExtraChar(self, s, dictionary):
        dp = [-1] * len(s)
        st = set(dictionary)
        return self.dfs(s, dp, st, 0)
        